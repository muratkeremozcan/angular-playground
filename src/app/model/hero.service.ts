import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders, HttpErrorResponse } from '@angular/common/http';
import { Observable } from 'rxjs';
import { catchError, map, tap } from 'rxjs/operators';
import { Hero } from './hero';

const httpOptions = {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  headers: new HttpHeaders({ 'Content-Type': 'application/json' })
};

@Injectable({
  providedIn: 'root'
})
export class HeroService {
  readonly heroesUrl = 'api/heroes'; // URL to web api

  constructor(private http: HttpClient) {}

  /** GET heroes from the server */
  getHeroes(): Observable<Hero[]> {
    return this.http.get<Hero[]>(this.heroesUrl).pipe(
      tap((heroes) => this.log(`fetched heroes`)),
      catchError(this.handleError('getHeroes'))
    ) as Observable<Hero[]>;
  }

  /** GET hero by id. Return `undefined` when id not found */
  getHero<Data>(id: number | string): Observable<Hero> {
    if (typeof id === 'string') {
      id = parseInt(id, 10);
    }
    const url = `${this.heroesUrl}/?id=${id}`;
    return this.http.get<Hero[]>(url).pipe(
      map((heroes) => heroes[0]), // returns a {0|1} element array
      tap((h) => {
        const outcome = h ? `fetched` : `did not find`;
        this.log(`${outcome} hero id=${id}`);
      }),
      catchError(this.handleError<Hero>(`getHero id=${id}`))
    );
  }

  //////// Save methods //////////

  /** POST: add a new hero to the server */
  addHero(hero: Hero): Observable<Hero> {
    return this.http.post<Hero>(this.heroesUrl, hero, httpOptions).pipe(
      tap((addedHero) => this.log(`added hero w/ id=${addedHero.id}`)),
      catchError(this.handleError<Hero>('addHero'))
    );
  }
  /** DELETE: delete the hero from the server */
  deleteHero(hero: Hero | number): Observable<Hero> {
    const id = typeof hero === 'number' ? hero : hero.id;
    const url = `${this.heroesUrl}/${id}`;

    return this.http.delete<Hero>(url, httpOptions).pipe(
      tap((_) => this.log(`deleted hero id=${id}`)),
      catchError(this.handleError<Hero>('deleteHero'))
    );
  }

  /** PUT: update the hero on the server */
  updateHero(hero: Hero): Observable<any> {
    return this.http.put(this.heroesUrl, hero, httpOptions).pipe(
      tap((_) => this.log(`updated hero id=${hero.id}`)),
      catchError(this.handleError<any>('updateHero'))
    );
  }
  /**
   * Returns a function that handles Http operation failures.
   * This error handler lets the app continue to run as if no error occurred.
   *
   * @param operation - name of the operation that failed
   */
  private handleError<T>(operation = 'operation') {
    return (error: HttpErrorResponse): Observable<T> => {
      const message =
        error.error instanceof ErrorEvent
          ? error.error.message
          : `server returned code ${error.status} with body "${error.error}"`;

      throw new Error(`${operation} failed: ${message}`);
    };
  }

  private log(message: string) {
    // console.log('HeroService: ' + message);
  }
}

/*
Copyright Google LLC. All Rights Reserved.
Use of this source code is governed by an MIT-style license that
can be found in the LICENSE file at https://angular.io/license
*/
