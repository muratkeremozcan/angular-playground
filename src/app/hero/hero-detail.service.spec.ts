import { HeroDetailService } from './hero-detail.service';
import { HeroService } from '../model';
import { SpectatorService, createServiceFactory } from '@ngneat/spectator/jest';
import { of } from 'rxjs';
import { fakeAsync, tick, waitForAsync } from '@angular/core/testing';

// [2].2 testing services  with other service dependencies
// (2.1) setup the service and satisfy the TS (2.2) use the service
// (2.3) use waitForAsync() or fakeAsync() pattern for testing promises or observables

// (2.4.0) KEY: with spectator, you can fully mock the dependency using the mocks property (no need for custom mocks unless you need them)
// (2.4.1) inject the service under test and the mock dependency
// (2.4.2) stub the external service's return value, and exercise the main service under test

describe('HeroDetailService', () => {
  let spectator: SpectatorService<HeroDetailService>;
  let heroDetailService: HeroDetailService;
  let heroServiceSpy: HeroService;

  const createService = createServiceFactory({
    service: HeroDetailService,
    // (2.4.0) with spectator, you can fully mock the dependency using the mocks property (no need for custom mocks unless you need them)
    mocks: [HeroService]
  });

  const stubHero = {
    id: 1,
    name: 'abc'
  };

  beforeEach(() => {
    // (2.1) (2.2) setup the service and satisfy the TS
    spectator = createService();
    heroDetailService = spectator.service;

    // (2.4.1) inject the service under test and the mock dependency
    heroDetailService = spectator.inject(HeroDetailService);
    heroServiceSpy = spectator.inject(HeroService);
  });

  it('sanity', () => {
    expect(heroDetailService).toBeTruthy();
    expect(heroDetailService).toMatchSnapshot();
  });

  describe('getHero', () => {
    it(
      'getHero with waitForAsync',
      waitForAsync(() => {
        // (2.4.2) stub the external service's return value, and exercise the main service under test
        jest.spyOn(heroServiceSpy, 'getHero').mockReturnValue(of(stubHero));

        // (2.3) use waitForAsync() or fakeAsync() pattern for testing promises or observables
        heroDetailService.getHero(1).subscribe((value) => {
          expect(heroServiceSpy.getHero).toHaveBeenCalledWith(1);
          expect(value).toEqual(stubHero);
          expect(value).toMatchSnapshot(); // this is a good example of using snapshot testing in services
        });
      })
    );

    it(
      'getHero with waitForAsync, cover the parseInt line',
      waitForAsync(() => {
        jest.spyOn(heroServiceSpy, 'getHero').mockReturnValue(of(stubHero));

        // to cover the source code line "if (typeof id === 'string') {" we pass in a string as the argument
        heroDetailService.getHero('1').subscribe((value) => {
          expect(heroServiceSpy.getHero).toHaveBeenCalledWith(1);
          expect(value).toEqual(stubHero);
          expect(value).toMatchSnapshot();
        });
      })
    );

    it('getHero with fakeAsync alternative', fakeAsync(() => {
      // (2.4.2) stub the external service's return value, and exercise the main service under test
      jest.spyOn(heroServiceSpy, 'getHero').mockReturnValue(of(stubHero));

      let expectedValue;
      // (2.3) use waitForAsync() or fakeAsync() pattern for testing promises or observables
      heroDetailService.getHero(1).subscribe((value) => (expectedValue = value));
      tick();
      expect(heroServiceSpy.getHero).toHaveBeenCalled();
      expect(expectedValue).toEqual(stubHero);
      expect(expectedValue).toMatchSnapshot();
    }));
  });

  describe('saveHero', () => {
    it(
      'saveHero with waitForAsync',
      waitForAsync(() => {
        // (2.4.2) stub the external service's return value, and exercise the main service under test
        jest.spyOn(heroServiceSpy, 'updateHero').mockReturnValue(of(null));

        // (2.3) use waitForAsync() or fakeAsync() pattern for testing promises or observables
        heroDetailService.saveHero(stubHero).subscribe((value) => {
          expect(heroServiceSpy.updateHero).toHaveBeenCalledWith(stubHero);
          expect(value).toBeNull();
          expect(value).toMatchSnapshot();
        });
      })
    );
    it('saveHero with fakeAsync', fakeAsync(() => {
      // (2.4.2) stub the external service's return value, and exercise the main service under test
      jest.spyOn(heroServiceSpy, 'updateHero').mockReturnValue(of(null));

      let expected;

      // (2.3) use waitForAsync() or fakeAsync() pattern for testing promises or observables
      heroDetailService.saveHero(stubHero).subscribe((value) => (expected = value));

      expect(heroServiceSpy.updateHero).toHaveBeenCalledWith(stubHero);
      expect(expected).toBeNull();
      expect(expected).toMatchSnapshot();
    }));
  });
});
