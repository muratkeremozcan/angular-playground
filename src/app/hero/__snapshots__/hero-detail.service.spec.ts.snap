// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroDetailService getHero getHero with fakeAsync alternative 1`] = `
{
  "id": 1,
  "name": "abc",
}
`;

exports[`HeroDetailService getHero getHero with waitForAsync 1`] = `
{
  "id": 1,
  "name": "abc",
}
`;

exports[`HeroDetailService getHero getHero with waitForAsync, cover the parseInt line 1`] = `
{
  "id": 1,
  "name": "abc",
}
`;

exports[`HeroDetailService sanity 1`] = `
HeroDetailService {
  "heroService": {
    "addHero": [MockFunction],
    "castToWritable": [Function],
    "deleteHero": [MockFunction],
    "getHero": [MockFunction],
    "getHeroes": [MockFunction],
    "handleError": [MockFunction],
    "log": [MockFunction],
    "updateHero": [MockFunction],
  },
}
`;

exports[`HeroDetailService saveHero saveHero with fakeAsync 1`] = `null`;

exports[`HeroDetailService saveHero saveHero with waitForAsync 1`] = `null`;
